src/Data/Variant.purs

```diff
@@ -73,6 +79,70 @@ on p f g r =
   coerceR ∷ Variant r2 → Variant r1
   coerceR = unsafeCoerce
 
+-- | Type class that matches a row for a `record` that will eliminate a row for
+-- | a `variant`, producing a `result` of the specified type. Just a wrapper for
+-- | `VariantMatch` to convert `RowToList` and vice versa.
+class VariantRecordElim
+    (variant ∷ # Type)
+    (record ∷ # Type)
+    result
+  | variant result → record
+  , record → variant result
+instance variantRecordElim
+  ∷ ( RowToList variant vlist
+    , RowToList record rlist
+    , VariantMatch vlist rlist result
+    , ListToRow vlist variant
+    , ListToRow rlist record )
```

Comment:
I think they are: in different situations either `variant` or `record` are being computed - you need something to produce what form they should take (neither `RowToList` nor `VariantMatch` do this).  If you only consider them both together as "input", then you won't need them.  But then your type inference will suck.